src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -986,151 +1075,255 @@ private BoundCollectionExpression ConvertCollectionExpression(
+                        // the parameterless `with()` is allowed.
+                        if (withElement.Arguments.Length > 0)
+                        {
+                            @this._diagnostics.Add(ErrorCode.ERR_CollectionArgumentsMustBeEmpty, withSyntax.GetFirstToken().GetLocation());
+                        }
+
+                        // Note: we intentionally report null here.  Even though the code has `with()` in it, we're not actually
+                        // going to call a particular constructor.  The lowering phase will properly handle creating a read-only
+                        // interface instance.
+                        return null;
+                    }
+
+                    var isMutableArray = @this._targetType.IsMutableArrayInterface(out var typeArgument);
+                    Debug.Assert(isMutableArray);
+
+                    // For the mutable array interfaces (ICollection<E>, IList<E>), we allow only the no-arg `with()` and
+                    // single int arg for the `List(int capacity)` case.
                     //
-                    // We could also consider removing this from the set of candidate methods when building them.
-                    collectionCreation = null;
-                    collectionBuilderMethod = null;
-                    collectionBuilderElementsPlaceholder = null;
+                    // This is from
+                    // https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#mutable-interface-translation
+                    // which dictates that IList<E> and ICollection<E> map to List<E>. and from
+                    // https://github.com/dotnet/csharplang/blob/90f1d8b0e9ba8a140f73aef376833969cce8bf9e/proposals/collection-expression-arguments.md?plain=1#L284
+                    // which dictates the two constructors we allow from that.
+
+                    var useSiteInfo = @this._binder.GetNewCompoundUseSiteInfo(@this._diagnostics);
+
+                    // Map to the corresponding List<E>() and List<E>(int) constructors in the constructed List<E> target.
+
+                    var constructedListType = @this._binder
+                        .GetWellKnownType(WellKnownType.System_Collections_Generic_List_T, ref useSiteInfo)
+                        .Construct([typeArgument]);
+
+                    // Don't need to report diagnostics here.  Our caller will have already done this.
+                    var list_T__ctor = (MethodSymbol?)@this._binder.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_List_T__ctor);
+                    var list_T__ctorInt32 = (MethodSymbol?)@this._binder.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_List_T__ctorInt32);
+
+                    var candidateConstructorsBuilder = ArrayBuilder<MethodSymbol>.GetInstance();
```

Comment:
The 99.9% case is that both of these symbols are available here, consider passing capacity `2` so that we may be able to move the resulting array from builder to ImmutableArray in many cases.